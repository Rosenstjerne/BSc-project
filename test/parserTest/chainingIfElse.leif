var int a,b;
a = 1;
b = 2;


if(a>b){
    print(1);
} else if(a<b){
    print(2);
} else if(a==b){
    print(3);
} else {
    print(4);
}