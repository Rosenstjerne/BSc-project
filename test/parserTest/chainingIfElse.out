
.data
mmapEr:
                .string "mmap failed\n" # error message string for mmap

form:
                .string "%d\n"          # form string for C printf

formTrue:
                .string "true\n"        # form string for C printf

formFalse:
                .string "false\n"       # form string for C printf

.text

.globl main

main:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $-16, %rsp         # allocate space for local variables


                movq %rsp, %rbx         # Saves stack pointer in %rbx
                addq $-72, %rsp         # allocate space for extra intermediate
                                        # stack memory

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save

                movq $0, -16(%rbp)      # Initializes b to 0
                movq $0, -8(%rbp)       # Initializes a to 0
                movq $1, %r9            # Moves integer into reg1
                movq %rbp, %rdx         # preparing for static link computation
                movq %r9, -8(%rdx)      # Move param reg1 into ['fun_1_main',
                                        # 'var_0_a'] (a)
                movq $2, %r11           # Moves integer into reg3
                movq %rbp, %rdx         # preparing for static link computation
                movq %r11, -16(%rdx)    # Move param reg3 into ['fun_1_main',
                                        # 'var_1_b'] (b)
                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %r12     # Move param var_0_a (a) into reg4
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %r13    # Move param var_1_b (b) into reg5
                movq %r13, %rcx         # Moves reg5 out into %rcx for cmp
                movq %r12, %rax         # Moves reg4 out into %rax for cmp
                cmpq %rcx, %rax         # eval reg4 > reg5
                jg lbl0_cmp_true        # Jump if the expression was true
                movq $0, %rax           # Moves false into %rax
                jmp lbl1_cmp_end        # Jump to end of expression
lbl0_cmp_true:
                movq $1, %rax           # Moves true into %rax
lbl1_cmp_end:
                movq %rax, %r14         # Moves reg6 out into %rax for cmp
                movq %r14, %rcx         # Moves reg6 out into %rax for
                                        # operation
                cmpq $1, %rcx           # Compare: reg6 == true
                jne lbl10_if_else_else  # Jump to else if false
                movq $1, %r15           # Moves integer into reg7
                movq %r15, %rcx         # Moves reg7 out into %rcx for print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
                jmp lbl11_if_else_end   # Jump to end
lbl10_if_else_else:
                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %rcx     # 
                movq %rcx, 0(%rbx)      # Move param var_0_a (a) into reg8
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %rcx    # 
                movq %rcx, -8(%rbx)     # Move param var_1_b (b) into reg9
                movq -8(%rbx), %rcx     # Moves reg9 out of the stack for cmp
                movq 0(%rbx), %rax      # Moves reg8 out of the stack for cmp
                cmpq %rcx, %rax         # eval reg8 < reg9
                jl lbl2_cmp_true        # Jump if the expression was true
                movq $0, %rax           # Moves false into %rax
                jmp lbl3_cmp_end        # Jump to end of expression
lbl2_cmp_true:
                movq $1, %rax           # Moves true into %rax
lbl3_cmp_end:
                movq %rax, -16(%rbx)    # Moves result into reg10
                movq -16(%rbx), %rcx    # Moves reg10 out of the stack for
                                        # operation
                cmpq $1, %rcx           # Compare: reg10 == true
                jne lbl8_if_else_else   # Jump to else if false
                movq $2, -24(%rbx)      # Moves integer into reg11
                movq -24(%rbx), %rcx    # Moves reg11 out of the stack for
                                        # print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
                jmp lbl9_if_else_end    # Jump to end
lbl8_if_else_else:
                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %rcx     # 
                movq %rcx, -32(%rbx)    # Move param var_0_a (a) into reg12
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %rcx    # 
                movq %rcx, -40(%rbx)    # Move param var_1_b (b) into reg13
                movq -40(%rbx), %rcx    # Moves reg13 out of the stack for cmp
                movq -32(%rbx), %rax    # Moves reg12 out of the stack for cmp
                cmpq %rcx, %rax         # eval reg12 == reg13
                je lbl4_cmp_true        # Jump if the expression was true
                movq $0, %rax           # Moves false into %rax
                jmp lbl5_cmp_end        # Jump to end of expression
lbl4_cmp_true:
                movq $1, %rax           # Moves true into %rax
lbl5_cmp_end:
                movq %rax, -48(%rbx)    # Moves result into reg14
                movq -48(%rbx), %rcx    # Moves reg14 out of the stack for
                                        # operation
                cmpq $1, %rcx           # Compare: reg14 == true
                jne lbl6_if_else_else   # Jump to else if false
                movq $3, -56(%rbx)      # Moves integer into reg15
                movq -56(%rbx), %rcx    # Moves reg15 out of the stack for
                                        # print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
                jmp lbl7_if_else_end    # Jump to end
lbl6_if_else_else:
                movq $4, -64(%rbx)      # Moves integer into reg16
                movq -64(%rbx), %rcx    # Moves reg16 out of the stack for
                                        # print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
lbl7_if_else_end:
lbl9_if_else_end:
lbl11_if_else_end:
end_main:

                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                addq $72, %rsp          # deallocate stack space for extra
                                        # intermediate stack memory


                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call

mmap_error:
                leaq mmapEr(%rip), %rdi # pass 1. arg in %rdi
                xor %rax, %rax          # no floating point arguments
                callq printf            # calls printf method

                movq $60, %rax          # prepare for exit syscall
                xor %rdi, %rdi          # exit status 0
                syscall                 # exit


